### PR TITLE
Taking the nthRoot of both sides

### PR DESCRIPTION
**Pull request by aliang8**
More support for finding roots of an equation.

e.g. `x^2 - 2 = 0` - > `x = [nthRoot(2, 2)]`
e.g. `(y + 1)^2 = 1` - > `y = 0`

### CODE REVIEW

File: lib/simplifyExpression/stepThrough.js [hunk 1]
```diff
@@ -84,10 +84,10 @@ function step(node) {
     breakUpNumeratorSearch,
     // e.g. 3/x * 2x/5 => (3 * 2x) / (x * 5)
     multiplyFractionsSearch,
-    // e.g. (2x + 3)(x + 4) => 2x^2 + 11x + 12
-    distributeSearch,
     // e.g. abs(-4) => 4
     functionsSearch,
+    // e.g. (2x + 3)(x + 4) => 2x^2 + 11x + 12
```
Comment (ldworkin) [inline lib/simplifyExpression/stepThrough.js]: Why this change of order?

File: lib/ChangeTypes.js [hunk 2]
```diff
@@ -180,6 +180,8 @@ module.exports = {
   SIMPLIFY_RIGHT_SIDE: 'SIMPLIFY_RIGHT_SIDE',
   // e.g. x + 3 = 2 -> x + 3 - 3 = 2 - 3
   SUBTRACT_FROM_BOTH_SIDES: 'SUBTRACT_FROM_BOTH_SIDES',
+  // e.g. x^2 = 1, nthRoot(x^2) = nthRoot(1)
+  NTH_ROOT_BOTH_SIDES: 'NTH_ROOT_BOTH_SIDES',
```
Comment (ldworkin) [inline lib/ChangeTypes.js]: Can we do +/- nthRoot(..)? What would we have to do to support +/-?

File: lib/solveEquation/EquationOperations.js [hunk 3]
```diff
@@ -125,13 +124,22 @@ EquationOperations.isolateSymbolOnLeftSide = function(equation, symbolName) {
   let nonSymbolTerm = Symbols.getLastNonSymbolTerm(leftNode, symbolName);
 
   let inverseOp, inverseTerm, changeType;
-  if (!nonSymbolTerm) {
+
+  // If there is no constant term on the left side of the equation
+  // and the left side is not a polynomialTerm
+  // or the left side is a symbol, return no change
+  // e.g. x = 0, no change
```
Comment (ldworkin) [inline lib/solveEquation/EquationOperations.js]: Anthony is the new docstring king 👑

File: lib/ChangeTypes.js (hunk 2 above)
Comment (aliang8) [inline lib/ChangeTypes.js]: We just have to display the +/- sign in latex :D

File: lib/solveEquation/EquationOperations.js [hunk 4]
```diff
@@ -125,13 +124,22 @@ EquationOperations.isolateSymbolOnLeftSide = function(equation, symbolName) {
...
+  // e.g. 2x + 3x = 1, no change
+  // e.g. x^2 = 1, change
+  if (!nonSymbolTerm && !Node.Type.isOperator(leftNode, '^')) {
     return EquationStatus.noChange(equation);
   }
 
   // Clone it so that any operations on it don't affect the node already
   // in the equation
-  nonSymbolTerm = clone(nonSymbolTerm);
+  nonSymbolTerm = nonSymbolTerm
```
Comment (ldworkin) [inline lib/solveEquation/EquationOperations.js]: Why do we need the ternary operator?

File: lib/solveEquation/EquationOperations.js [hunk 5]
```diff
@@ -177,8 +185,9 @@ EquationOperations.isolateSymbolOnLeftSide = function(equation, symbolName) {
       }
     }
     else if (leftNode.op === '^') {
-      // TODO: support roots
```
Comment (ldworkin) [inline lib/solveEquation/EquationOperations.js]: Loving killing TODOs! 🎉

File: lib/solveEquation/EquationOperations.js [hunk 6]
```diff
@@ -203,11 +212,21 @@ function performTermOperationOnEquation(equation, op, term, changeType) {
   const oldEquation = equation.clone();
 
   const leftTerm = clone(term);
-  const rightTerm = clone(term);
+
+  // The root is the exponent of the polynomial term on the left side
+  const root = op === 'nthRoot'
+        ? leftTerm.args[1]
```
Comment (ldworkin) [inline lib/solveEquation/EquationOperations.js]: Use the getRoot helper function!!!!

**Comment by ldworkin** [comment]
Just did a very cursory pass, will come back and re-read later today, but looks great!

File: lib/ChangeTypes.js (hunk 2 above)
Comment (ldworkin) [inline lib/ChangeTypes.js]: Yup there's a latex sign for that! /pm I think

Comment (ldworkin) [inline lib/ChangeTypes.js]: Hmm I actually think the plus/minus might be more complicated ... I ran into it when I was thinking about using the quadratic equation also. It might be tricky cause mathjs doesn't support it. I'm thinking maybe we could use a custom function? Let's talk about it.

File: test/solveEquation/solveEquation.test.js [hunk 7]
```diff
@@ -88,17 +86,32 @@ describe('solveEquation for =', function () {
...
+    ['x^2 = 1', 'x = 1'],
+    ['x^3 - 3 = 0', 'x = nthRoot(3, 3)'],
+    ['(x^2 - 2) (x^2 - 5) = 0', 'x = [nthRoot(2, 2), nthRoot(5, 2)]'],
+    ['(x^2 + 2) (x^3 - 7) = 0', 'x = [nthRoot(-2, 2), nthRoot(7, 3)]'],
+    ['x^2 + 1 = 0', 'x = nthRoot(-1, 2)'],
+    ['(y + 1)^2 = 1', 'y = 0'],
+    ['(y + 1)^3 = 8', 'y = 1'],
+    ['(2x + 1)^3 = 1', 'x = 0'],
+    ['(3x + 2)^2 = 2', 'x = nthRoot(2, 2) / 3 - 2/3'],
+    // TODO: fix this
```
Comment (ldworkin) [inline test/solveEquation/solveEquation.test.js]: Yea, looks like it fails due to complex #, so should just be moved below

File: test/solveEquation/solveEquation.test.js [hunk 8]
```diff
@@ -88,17 +86,32 @@ describe('solveEquation for =', function () {
...
+    // ['(3x + 2)^2 + 2 = 1']
     // TODO: fix these cases, fail because lack of factoring support, for complex #s,
     // for taking the sqrt of both sides, etc
```
Comment (ldworkin) [inline test/solveEquation/solveEquation.test.js]: This comment is no longer right, because we do take the square root of both sides now, so any cases that were previously failing because of that shouldn't be anymore (and should be uncommented out)

File: lib/simplifyExpression/stepThrough.js (hunk 1 above)
Comment (aliang8) [inline lib/simplifyExpression/stepThrough.js]: We want to evaluate the multiplication of roots before distributing.

Comment (ldworkin) [inline lib/simplifyExpression/stepThrough.js]: Can you give an example so I can see what you mean by this / why it's important?

Also, just thought of something: Does this PR conflict with your PR to expand exponents? E.g. for `(y + 1)^2 = 1`, do we want `y+1 = sqrt(1)` or `(y+1)(y+1) = 1`?
